PS-1525: Adding the innodb_encrypt_tables system variable.

dutow opened the pull request:
Possible values: ON, OFF, FORCE.

If an InnoDb table is created with this set to ON or FORCE without
any encryption setting specified, it will be created encrypted, and
the encryption setting will show up in it's creation info.

dutow commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -556,11 +556,11 @@ ha_innobase::check_if_supported_inplace_alter(
 
-	if (Encryption::is_none(old_encryption)
-	    != Encryption::is_none(new_encryption)) {
+	if (currently_encrypted != !Encryption::is_none(new_encryption)) {
```
I'm quite sure that I'll have to add another check here: with this, currently we allow inplace alterations for non encrypted tables, even when innodb_encrypt_tables is FORCE. This is consistent with mariadb, but I don't think it's logical. Currently waiting for some opinions about this.

laurynas-biveinis replied:
Consider "==" instead of "!= !"

laurynas-biveinis commented on mysql-test/suite/innodb/t/encryption_force.test:
```diff
@@ -0,0 +1,107 @@
+# Restart the server with keyring loaded
+--let $restart_parameters="$keyring2_restart_param"
+--let $restart_hide_args= 1
+--source include/restart_mysqld.inc
```
This is the sole restart in the testcase - can it be performed by the testcase options file?

laurynas-biveinis commented on mysql-test/suite/innodb/t/encryption_force.test:
```diff
@@ -0,0 +1,107 @@
+-- source include/have_innodb.inc
```
I'd name "percona_force_encryption" so that it's immediately visible in the runs as a Percona feature testcase

laurynas-biveinis commented on mysql-test/suite/sys_vars/r/all_vars.result:
```diff
@@ -15,6 +15,8 @@ left join t1 on variable_name=test_name where test_name is null ORDER BY variabl
 There should be *no* variables listed below:
 DISABLED_STORAGE_ENGINES
 DISABLED_STORAGE_ENGINES
+INNODB_ENCRYPT_TABLES
```
Please review this testcase, specifically the "There should be *no* variables listed below:" comment above :)

As you create sys_vars.innodb_encrypt_tables_basic, move the sysvar tests from the feature testcase

laurynas-biveinis commented on mysql-test/suite/innodb/t/encryption_force.test:
```diff
@@ -0,0 +1,107 @@
+insert into t6_force_default values (42);
+
+--error ER_WRONG_VALUE_FOR_VAR
+set global innodb_encrypt_tables='FOO';
```
Move this elsewhere (see the comment at sys_vars.all_vars.result)

laurynas-biveinis commented on mysql-test/suite/innodb/t/encryption_force.test:
```diff
@@ -0,0 +1,107 @@
+alter table t4_on_explicit_off encryption="N";
+
+set global innodb_encrypt_tables='ON';
+--error 1846
```
s/1846/ER_something/g

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -2681,6 +2681,22 @@ Encryption::is_none(const char* algorithm)
+	/* NULL is the same as NONE */
+	if (algorithm != NULL && (
+	    innobase_strcasecmp(algorithm, "n") == 0
+	    || innobase_strcasecmp(algorithm, "") == 0)) {
```
maybe just algorithm[0] == '\0'

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11080,7 +11096,11 @@ create_table_info_t::create_table_def()
 
-		if (!Encryption::is_none(encrypt)) {
+		const bool default_to_encrypt =
+			!Encryption::none_specified(encrypt)
```
For reader very hard to tell the difference between "none_specified" and "is_none", consider renaming

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11727,6 +11747,14 @@ create_table_info_t::create_option_encryption_is_valid() const
 	bool table_is_encrypted =
 		!Encryption::is_none(m_create_info->encrypt_type.str);
 
+	if(srv_encrypt_tables == 2
```
Space after if

laurynas-biveinis replied:
Introduce enum for 0, 1, 2 here

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11989,6 +12017,11 @@ ha_innobase::update_create_info(
 	    && create_info->tablespace == NULL) {
 		create_info->tablespace = m_prebuilt->table->tablespace;
 	}
+	if((m_prebuilt->table->flags2 & DICT_TF2_ENCRYPTION)
```
Space after if

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21005,6 +21038,19 @@ static MYSQL_SYSVAR_ULONG(autoextend_increment,
+};
+static MYSQL_SYSVAR_ENUM(encrypt_tables, srv_encrypt_tables,
+			 PLUGIN_VAR_OPCMDARG,
+			 "Enable encryption for tables. ",
```
Describe the effect of the three values

dutow commented:
Everything's fixed, I just tried to do a successful jenkins run... but wasn't able, as usual for 5.7.

So this is it: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1503/

Most platforms have warnings/network errors, bun two builds succeeded without any warnings.

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,96 @@
+insert into t5_off_explicit_off values (42);
+insert into t6_force_default values (42);
+
+--error ER_WRONG_VALUE_FOR_VAR
```
Remove, now covered by the sys_vars test

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,96 @@
+alter table t3_on_default encryption="N";
+alter table t4_on_explicit_off ALGORITHM=INPLACE;
+alter table t4_on_explicit_off encryption="N";
+--error 1846
```
s/1846/ER_...

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,96 @@
+drop table t5_off_explicit_off;
+drop table t6_force_default;
+
+#Cleanup
```
Space after #

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -2681,6 +2681,21 @@ Encryption::is_none(const char* algorithm)
 	return(false);
 }
 
+/** Check if the NO algorith was explicitly specified.
```
s/algorith/algorithm

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -2681,6 +2681,21 @@ Encryption::is_none(const char* algorithm)
+Encryption::none_explicitly_specified(const char* algorithm)
+{
+	/* NULL is the same as NONE */
+	if (algorithm != NULL && 
```
{
return (algorithm != NULL && innobase_strcmp() == 0);
}

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -448,12 +448,18 @@ struct Encryption {
 	static bool is_none(const char* algorithm)
 		MY_ATTRIBUTE((warn_unused_result));
 
+	/** Check if the NO algorithm was explicitly speficied.
```
"specified"

dutow commented:
New jenkins run: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1510/

I also deleted a comment from `none_explicitly_specified`, which was misleading (I copied that from `is_none`, but it wasn't true for the new function)

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,93 @@
+-- source include/have_innodb.inc
```
If you have to re-push due to other review comments, then no space between -- and source

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,93 @@
+-- source include/have_innodb.inc
+-- source include/not_embedded.inc
+-- source include/have_partition.inc
```
Partitioning is not used in the testcase - adding such tests might be a good idea

laurynas-biveinis commented on mysql-test/suite/sys_vars/t/innodb_encrypt_tables_basic.test:
```diff
@@ -0,0 +1,27 @@
+
+# If we are lacking permissions, then we cannot restore the startup value
+# at the end.
+--source include/restart_mysqld.inc
```
? If we are lacking permissions, then we could not switch away from the startup value neither. And how can the MTR MySQL user lack permissions?

dutow replied:
I based this on another testcase, and this was included in it - I assumed it made sense. I'll remove it then.

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11989,6 +12018,11 @@ ha_innobase::update_create_info(
+	if ((m_prebuilt->table->flags2 & DICT_TF2_ENCRYPTION)
+			&& create_info->encrypt_type.length == 0) {
+		static const LEX_STRING yes = { C_STRING_WITH_LEN("Y") };
+		create_info->encrypt_type = yes;
```
Use of stack variable after scope? Just in case, please grep, such LEX_STRING/LEX_CSTRING might be already available elsewhere for reuse

dutow replied:
I don't see a stack variable? I only see a static const.
But it would also work with a stack variable, because the assignment calls the copy constructor, and the struct only contains a const char pointer to a string literal, and an integer (the length)

laurynas-biveinis replied:
Right, sorry

dutow commented:
Comments fixed.
New jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1530/

percona-ysorokin commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,92 @@
+drop table t6_force_default;
+
+# Cleanup
+set global innodb_encrypt_tables='OFF';
```
save the value of the `@@global.innodb_encrypt_tables` system variable at the beginning of the test and restore it here.

laurynas-biveinis replied:
Then probably "select @@innodb_encrypt_tables;" at line 10 is redundant too

percona-ysorokin commented on mysql-test/suite/perfschema/r/show_sanity.result:
```diff
@@ -412,6 +412,7 @@ SHOW_MODE	SOURCE	VARIABLE_NAME
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_THRESHOLD
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_ZIP_LEVEL
 5.6	I_S.SESSION_VARIABLES	INNODB_DEADLOCK_DETECT
+5.6	I_S.SESSION_VARIABLES	INNODB_ENCRYPT_TABLES
```
The proper way to fix this is to extend `show_sanity.test` itself with a new known global variable rather than re-recording `.result` file.

percona-ysorokin commented on mysql-test/suite/perfschema/r/show_sanity.result:
```diff
@@ -441,6 +442,7 @@ SHOW_MODE	SOURCE	VARIABLE_NAME
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_THRESHOLD
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_ZIP_LEVEL
 5.6	I_S.SESSION_VARIABLES	INNODB_DEADLOCK_DETECT
+5.6	I_S.SESSION_VARIABLES	INNODB_ENCRYPT_TABLES
```
Likewise

percona-ysorokin commented on sql/sql_show.cc:
```diff
@@ -1923,7 +1923,11 @@ int store_create_info(THD *thd, TABLE_LIST *table_list, String *packet,
     {
       packet->append(STRING_WITH_LEN(" ENCRYPTION="));
-      append_unescaped(packet, share->encrypt_type.str, share->encrypt_type.length);
+      append_unescaped(packet, table->s->encrypt_type.str, table->s->encrypt_type.length);
```
Isn't `share` is just a synonym for `table->s` here?

dutow replied:
It is , but the condition used the other name - I thought it would be clearer to use the same on both places.

percona-ysorokin commented on sql/sql_show.cc:
```diff
@@ -1923,7 +1923,11 @@ int store_create_info(THD *thd, TABLE_LIST *table_list, String *packet,
       packet->append(STRING_WITH_LEN(" ENCRYPTION="));
-      append_unescaped(packet, share->encrypt_type.str, share->encrypt_type.length);
+      append_unescaped(packet, table->s->encrypt_type.str, table->s->encrypt_type.length);
+    } else if (create_info.encrypt_type.length)
```
Why do we need another branch here? Shouldn't `table->s->encrypt_type` be always populated with actual encryption type value?

dutow replied:
It's... similar, but not exactly the same. table->s is based on the sql parser. If the option was specified in the SQL query, it is in it.

But if a table is encrypted because of the variable, only the innodb flag is set, the information during the parsing is unchanged.

I found no good way to change it, but the `create_info` structure can be altered by the virtual `update_create_info` function, which I modified for innodb.

percona-ysorokin replied:
Still not sure why we have unsynchronized data here. A handler for `SHOW CREATE TABLE` (`mysqld_show_create()`) before calling `store_create_info()` function calls `open_tables()` - a function which should populate all the table attributes (including encryption flag). This `open_tables()` is used all over the code (not only for `SHOW CREATE TABLE`) and if it doesn't do this, we have a problem.

dutow commented:
Test cases updated.

laurynas-biveinis commented on mysql-test/suite/perfschema/t/show_sanity.test:
```diff
@@ -760,7 +760,8 @@ insert into test.sanity values
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "VERSION_COMPILE_MACHINE"),
   ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "VERSION_COMPILE_OS"),
-  ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "REQUIRE_SECURE_TRANSPORT")
+  ("JUNK: GLOBAL-ONLY", "I_S.SESSION_VARIABLES", "REQUIRE_SECURE_TRANSPORT"),
```
This list is maintained sorted

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21005,6 +21039,22 @@ static MYSQL_SYSVAR_ULONG(autoextend_increment,
+static MYSQL_SYSVAR_ENUM(encrypt_tables, srv_encrypt_tables,
+			 PLUGIN_VAR_OPCMDARG,
+			 "Enable encryption for tables. "
+			 "When turned ON, the default encryption setting is yes when it isn't "
```
Maybe "When turned ON, all tables are created encrypted unless specified otherwise."

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21005,6 +21039,22 @@ static MYSQL_SYSVAR_ULONG(autoextend_increment,
+			 "Enable encryption for tables. "
+			 "When turned ON, the default encryption setting is yes when it isn't "
+			 "specified. When it's set to FORCE, only encrypted tables can be created."
+			 "The FORCE setting also disables non inplace alteration of unencryted tables,"
```
"unencryted"
Also maybe "The FORCE setting also disables ALTER TABLE operations for unencrypted tables that would cause table rebuilds"

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -21005,6 +21039,22 @@ static MYSQL_SYSVAR_ULONG(autoextend_increment,
+			 "When turned ON, the default encryption setting is yes when it isn't "
+			 "specified. When it's set to FORCE, only encrypted tables can be created."
+			 "The FORCE setting also disables non inplace alteration of unencryted tables,"
+			 " without encryting them in the process.",
```
"encryting"

percona-ysorokin commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,92 @@
+drop table t4_on_explicit_off;
+drop table t5_off_explicit_off;
+drop table t6_force_default;
+
```
Please, also add tests for how `CREATE TABLE LIKE <existing_table>` works for different combinations of `innodb_encrypt_tables` and encryption attribute of `<existing_table>`.

dutow replied:
I added two new tables, their name ending with _dup

laurynas-biveinis replied:
There are no DROP TABLE for _dup tables?

percona-ysorokin commented:
Another big question is whether `ENCRYPTION` attribute should be replicated depending on the value of `innodb_encrypt_tables` on the slave?

dutow commented:
Updated PR:
* added the adjust_create_info_for_frm function to the handles class. This function allows any storage engine to modify the table definition before it's written into the frm file
* rebased to latest 5.7 head

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,96 @@
+--source include/have_innodb.inc
```
Since the feature goes through .FRM, its cold reading should be tested: add FLUSH TABLES, query the same tables again

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11958,6 +11983,20 @@ create_table_info_t::create_options_are_invalid()
+void
+ha_innobase::adjust_create_info_for_frm(
+/*============================*/
+	HA_CREATE_INFO*	create_info)	/*!< in/out: create info */
```
New-InnoDB doxygen style header comment

dutow commented:
PR updated.

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_force_encryption.test:
```diff
@@ -0,0 +1,111 @@
+
+let $initial_encrypt_tables_value = `select @@innodb_encrypt_tables`;
+
+create table t1_default_explicit_on (a int) engine=innodb encryption="Y";
```
I'm sorry if this was discussed already, but what about partitioned tables?

dutow replied:
They should work the same way as when you specify the encrypt flag manually - this change only adds the encrypt flag to the create table statement if it's not present, and reports an error if it is no and we want to farce encryption.

I'll add a few tests with them.

dutow replied:
Added test cases with partitions / tablespaces.

dutow commented:
Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1612/

percona-ysorokin commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11958,6 +11983,19 @@ create_table_info_t::create_options_are_invalid()
+{
+		if (create_info->encrypt_type.length == 0
+			&& create_info->encrypt_type.str == NULL
+			&& srv_encrypt_tables != 0
```
`srv_encrypt_tables != SRV_ENCRYPT_TABLES_OFF`

percona-ysorokin commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -11958,6 +11983,19 @@ create_table_info_t::create_options_are_invalid()
+			&& create_info->encrypt_type.str == NULL
+			&& srv_encrypt_tables != 0
+			) {
+				static const LEX_STRING yes = { C_STRING_WITH_LEN("Y") };
```
Function-level static variables/constants are not potentially thread-safe in c++98 (even for POD data types). Please, move to the global level or class level
`/*static*/ const LEX_STRING ha_innobase::yes = { C_STRING_WITH_LEN("Y") };`

dutow commented:
Updated.

dutow commented:
Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1646/

The centos6 build failed because there was no space on the disk.
